src/core/PlaneWaveBasis.jl

```diff
@@ -0,0 +1,216 @@
+                   wave vector. The defines the wave function basis ``B_{Ψ,k}`` for
+                   each ``k``-Point.
+- `FFT` and `iFFT`:   Containers for the forward and backward FFT plans on ``B_ρ``.
+"""
+struct PlaneWaveBasis{T <: Real}
+    lattice::SMatrix{3, 3, T}
+    recip_lattice::SMatrix{3, 3, T}
+    unit_cell_volume::T
+    recip_cell_volume::T
+
+    grid_size::SVector{3, Int}
+    Ecut::T
+    kpoints::Vector{SVector{3, T}}
+    kweights::Vector{T}
+    kbasis::Vector{Vector{SVector{3, Int}}}
+
+    FFT
+    iFFT
+end
+
+function PlaneWaveBasis(lattice::AbstractMatrix{T}, grid_size,
+                        Ecut::Number, kpoints, kweights) where {T <: Real}
+    lattice = SMatrix{3, 3, T}(lattice)
+    recip_lattice = 2π * inv(lattice')
+
+    # Plan a FFT, spending some time on finding an optimal algorithm
+    # for the machine on which the computation runs
+    grid_size = SVector{3, Int}(grid_size)
+    tmp = Array{Complex{T}}(undef, grid_size...)
+    fft_plan = plan_fft!(tmp, flags=FFTW.MEASURE)
+    ifft_plan = plan_ifft!(tmp, flags=FFTW.MEASURE)
+
+    dummy_kpoints = Vector{SVector{3, T}}(undef, length(kpoints))
+    dummy_kweights = Vector{T}(undef, length(kweights))
+    dummy_kbasis = Vector{Vector{SVector{3, Int}}}(undef, length(kpoints))
+    pw = PlaneWaveBasis{T}(lattice, recip_lattice, det(lattice), det(recip_lattice),
+                           grid_size, Ecut, dummy_kpoints, dummy_kweights, dummy_kbasis,
+                           fft_plan, ifft_plan)
+    substitute_kpoints!(pw, kpoints, kweights)
+end
+
+"""
+Replace the kpoints of an existing Plane-wave basis and make
+the internal data structures consistent.
+"""
+function substitute_kpoints!(pw::PlaneWaveBasis{T}, kpoints, kweights) where T
+    if length(kpoints) != length(kweights)
+        error("Lengths of kpoints and length of kweights need to agree")
+    end
+    kweights = Vector{T}(kweights / sum(kweights))  # Normalise kweights
+    kpoints = [SVector{3, T}(kp) for kp in kpoints]
+
+    resize!(pw.kpoints, length(kpoints))[:] = kpoints
+    resize!(pw.kweights, length(kweights))[:] = kweights
+
+    # Update kbasis: For each k-Point select those Gcoords, such that
+    # the kinetic energy cutoff is satisfied
+    for (ik, k) in enumerate(kpoints)
+        energy(Gc) = sum(abs2, pw.recip_lattice * (k + Gc)) / 2
+        p = [Gc for Gc in gcoords(pw) if energy(Gc) ≤ pw.Ecut]
+        pw.kbasis[ik] = [Gc for Gc in gcoords(pw) if energy(Gc) ≤ pw.Ecut]
+    end
+
+    pw
+end
+
+Base.eltype(basis::PlaneWaveBasis{T}) where T = T
+
+"""
+Return a generator producing the range of wave-vector coordinates contained
+in the Fourier grid ``B_ρ`` described by the plane-wave basis.
+"""
+function gcoords(pw::PlaneWaveBasis)
+    start = -ceil.(Int, (pw.grid_size .- 1) ./ 2)
+    stop  = floor.(Int, (pw.grid_size .- 1) ./ 2)
+    ci = CartesianIndices((UnitRange.(start, stop)..., ))
+    (SVector(ci[i].I) for i in 1:length(ci))
+end
```

Comment:
Can you add an `@assert` that the `grid_size` is sufficient wrt Ecut here?

Reply:
Why here? For me this is just a fast helper function to get the wave vectors, whenever they are explicitly needed. If one wanted to check for `Ecut` and for internal consistency of `PlaneWaveBasis`, I would put it into the constructor instead. Don't you think?